### PR TITLE
Support arbitrary `Optimizable`s as optimizers

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

This allows us to support works like [Gradient Descent: The Ultimate Optimizer](https://openreview.net/forum?id=-Qp-3L-5ZdI) ([code](https://github.com/kach/gradient-descent-the-ultimate-optimizer))

Example script:

```python
import pytorch_lightning as L
import torch
from pytorch_lightning.demos.boring_classes import RandomDataset

# pip install git+https://github.com/kach/gradient-descent-the-ultimate-optimizer
from gradient_descent_the_ultimate_optimizer import gdtuo


# Versions compatible with the requirements of torch optimizers
# Ideally these would be upstreamed so they are not required
class CompatibleAdam(gdtuo.Adam):
    param_groups = None
    defaults = None

    @property
    def state(self):
        return self.cache

    def state_dict(self):
        return {}

    def load_state_dict(self):
        pass


class CompatibleModuleWrapper(gdtuo.ModuleWrapper):
    param_groups = None
    defaults = None

    @property
    def state(self):
        return self.optimizer.state

    def step(self, closure):
        # ignore the closure - optional during manual optimization anyways
        return super().step()

    def state_dict(self):
        return self.optimizer.state_dict()

    def load_state_dict(self):
        return self.optimizer.load_state_dict()


class GDTUOLightningModule(L.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.automatic_optimization = False

    def configure_optimizers(self):
        optimizer0 = gdtuo.SGD(1e-5)
        optimizer1 = CompatibleAdam(optimizer=optimizer0)
        mw = CompatibleModuleWrapper(self, optimizer=optimizer1)
        mw.initialize()
        return mw

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch):
        mw = self.optimizers()
        mw.begin()  # call this before each step, enables gradient tracking on desired params
        mw.zero_grad()

        pred = self(batch)
        loss = torch.nn.functional.mse_loss(pred, torch.ones_like(pred))

        loss.backward(create_graph=True)  # important! use create_graph=True
        mw.step()

        return loss


dataset = RandomDataset(32, 64)
train_dataloader = torch.utils.data.DataLoader(dataset, batch_size=2)
trainer = L.Trainer(max_epochs=10, accelerator="auto")
model = GDTUOLightningModule()
trainer.fit(model, train_dataloader)
```

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda

### CODE REVIEW

**Comment by carmocca** [comment]
Hi paper authors @kach @0aax!

Would you accept a PR that adds the `state`, `state_dict`, ... attributes to the `gdtuo.Optimizable`s so that users don't have to do it? (see my script in the PR description). 

Sorry to ask here, the https://github.com/kach/gradient-descent-the-ultimate-optimizer repository does not have an issues tab.

**Review by awaelchli — APPROVED** [review]
@carmocca Anything that needs to be done for Lite? We don't have instance checks, but maybe the types?

**Review by Borda — APPROVED** [review]

**Comment by kach** [comment]
Hi @carmocca - we plan to leave the github repository as it is so that it remains a reproducible archive of the paper (hence no "issues" tab). Because the codebase is MIT-licensed, you are welcome to bundle a modified version with your library as long as you preserve the attribution notice.

**Comment by franchesoni** [comment]
@williamFalcon @justusschock 
sorry for spamming, but this feature would be great!

**Review by justusschock — APPROVED** [review]
